In [1]:
import neural_net_without_autograd as NNC
import numpy as np
import matplotlib.pyplot as plt

from sklearn import datasets
from sklearn.metrics import accuracy_score
from sklearn.model_selection import train_test_split

This file uses our neural network implementation to classify the MNIST digits dataset as a check of the performance of oyr model.

In [4]:
# ensure the same random numbers appear every time
np.random.seed(0)

digits = datasets.load_digits()

# define inputs and labels
inputs = digits.images
labels = digits.target

In [3]:
X_train, X_test, y_train, y_test = train_test_split(inputs, labels, test_size=0.2)

network_input_size = X_train.shape[1]
layer_output_sizes = [50, 10]
activations = ['sigmoid', 'softmax']

nn = NNC.NeuralNet(network_input_size, 
                layer_output_sizes, 
                activations, 
                loss_fn='cross_entropy', 
                epsilon=0.01, 
                epochs=100, 
                batch_size=16)
nn.fit(X_train, y_train)

predictions_train = nn.predict(X_train)
predictions_test = nn.predict(X_test)
print(f'Train accuracy: {accuracy_score(y_train, predictions_train)}')
print(f'Test accuracy: {accuracy_score(y_test, predictions_test)}')

print(nn.score(X_test, y_test))

Epoch 0: Accuracy = 0.6457898399443285
Train accuracy: 0.9993041057759221
Test accuracy: 0.9444444444444444
0.9444444444444444


With 100 epochs and non-optimized parameters we're able to achive a test score of 0.944, which is a good sign.